File: infrastructure/db-core/src/main/java/com/nmh/commerce/coupon/ConstraintPolicyEntity.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.BaseEntity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

@Entity
@DiscriminatorColumn(name = "constraint_policy_type")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
public abstract class ConstraintPolicyEntity extends BaseEntity {

    @ManyToOne
    @JoinColumn(name = "coupon_id")
    private CouponEntity coupon;

    protected ConstraintPolicyEntity(CouponEntity coupon) {
        this.coupon = coupon;
        coupon.addConstraintPolicy(this);
    }

    public static ConstraintPolicyEntity from(ConstraintPolicy policy, CouponEntity coupon) {
        if (policy instanceof MinimumPriceConstraintPolicy minimumPriceConstraintPolicy) {
            return MinimumPriceConstraintPolicyEntity.from(minimumPriceConstraintPolicy, coupon);
        }
        throw new IllegalArgumentException("지원하지 않는 제약 정책입니다: " + policy.getClass());
    }

    public abstract ConstraintPolicy toDomain();
}

File: infrastructure/db-core/src/main/kotlin/com/nmh/commerce/coupon/ConstraintPolicyEntity.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.BaseEntity
import jakarta.persistence.*

@Entity
@DiscriminatorColumn(name = "constraint_policy_type")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
abstract class ConstraintPolicyEntity protected constructor(
    override val id: Long = 0,
    @ManyToOne
    @JoinColumn(name = "coupon_id")
    private val coupon: CouponEntity
) : BaseEntity<Long>() {

    init {
        coupon.addConstraintPolicy(this)
    }

    abstract fun toDomain(): ConstraintPolicy

    companion object {
        fun from(policy: ConstraintPolicy, coupon: CouponEntity): ConstraintPolicyEntity {
            if (policy is MinimumPriceConstraintPolicy) {
                return MinimumPriceConstraintPolicyEntity.from(policy, coupon)
            }
            throw IllegalArgumentException("지원하지 않는 제약 정책입니다: " + policy.javaClass)
        }
    }
}
